name: Run Tests and Close PR on Failure

on:
  pull_request:
    types: [opened, synchronize, reopened]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Execute tests
      run: npm test

    - name: if fail
      if: failure()
      uses: actions/github-script@v6
      with:
        github-token: ${{ secrets.GITHUB_TOKEN }}
        script: |
          const ref = context.payload.pull_request.head.ref
          const pull_number = context.payload.pull_request.number

          await github.pulls.createReview({
            owner: context.repo.owner,
            repo: context.repo.repo,
            pull_number: pull_number,
            body: "테스트가 실패했습니다.",
            event: "REQUEST_CHANGES"
          })

          await github.pulls.update({
            owner: context.repo.owner,
            repo: context.repo.repo,
            pull_number: pull_number,
            state: "closed"
          })
